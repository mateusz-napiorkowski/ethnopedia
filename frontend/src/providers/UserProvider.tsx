import React, { createContext, useContext, useState, useEffect } from "react"
import { jwtDecode } from "jwt-decode"

export interface JWT {
    exp: number;
    username: string;
    firstName: string;
    userId: string;
}

interface UserContextProps {
    isUserLoggedIn: boolean;
    firstName: string;
    userId: string;
    jwtToken: string;
    setUserData: (isLoggedIn: boolean, name: string, jwtToken: string, userId: string) => void;
}

export const UserContext = createContext<UserContextProps>({
    isUserLoggedIn: false,
    firstName: "",
    userId: "",
    jwtToken: "",
    setUserData: () => {
    },
})

export const useUser = () => useContext(UserContext)

interface UserProviderProps {
    children: React.ReactNode;
}

export const UserProvider: React.FC<UserProviderProps> = ({ children }) => {
    const [isUserLoggedIn, setIsUserLoggedIn] = useState(false)
    const [firstName, setFirstName] = useState("")
    const [userId, setUserId] = useState("")
    const [jwtToken, setJwtToken] = useState("")

    const setUserData = (isLoggedIn: boolean, name: string, jwtToken: string, userId: string) => {
        setIsUserLoggedIn(isLoggedIn)
        setFirstName(name)
        setUserId(userId)
        setJwtToken(jwtToken)
    }

    useEffect(() => {
        const token = localStorage.getItem("token")

        if (token) {
            try {
                const decodedToken = jwtDecode<JWT>(token)
                const currentTime = Date.now() / 1000
                if (decodedToken.exp > currentTime) {
                    setUserData(true, decodedToken.firstName, token, decodedToken.userId)
                } else {
                    localStorage.removeItem("token")
                    setUserData(false, "", "", "")
                }
            } catch (error) {
                console.error("Invalid token:", error)
                localStorage.removeItem("token")
                setUserData(false, "", "", "")
            }
        }
    }, [])

    return (
        <UserContext.Provider value={{ isUserLoggedIn, firstName, userId, jwtToken, setUserData }}>
            {children}
        </UserContext.Provider>
    )
}
