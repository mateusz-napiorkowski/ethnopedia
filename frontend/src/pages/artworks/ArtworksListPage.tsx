import { useMutation, useQuery, useQueryClient } from "react-query";
import { getArtworksForCollectionPage, deleteArtworks } from "../../api/artworks";
import { getCollection } from "../../api/collections";
import LoadingPage from "../LoadingPage";
import { useEffect, useState } from "react";
import Navbar from "../../components/navbar/Navbar";
import { useLocation, useNavigate, useParams } from "react-router-dom";
import SearchComponent from "../../components/search/SearchComponent";
import ImportOptions from "../../components/ImportOptions";
import ExportOptions from "../../components/ExportOptions";
import { ReactComponent as PlusIcon } from "../../assets/icons/plus.svg";
import { ReactComponent as FileImportIcon } from "../../assets/icons/fileImport.svg";
import { ReactComponent as FileExportIcon } from "../../assets/icons/fileExport.svg";
import WarningPopup from "../WarningPopup";
import SortOptions from "../../components/SortOptions";
import Navigation from "../../components/Navigation";
import Pagination from "../../components/Pagination";
import { useUser } from "../../providers/UserProvider";
import { getAllCategories } from "../../api/categories";
import DisplayCategoriesSelect from "../../components/DisplayCategoriesSelect";
import { ReactComponent as EditIcon } from "../../assets/icons/edit.svg"
import ArtworksList from '../../components/artwork/ArtworksList';



const ArtworksListPage = ({ pageSize = 10 }) => {
    const [selectedArtworks, setSelectedArtworks] = useState<{ [key: string]: boolean }>({});
    const [showImportOptions, setShowImportOptions] = useState<boolean>(false);
    const [showExportOptions, setShowExportOptions] = useState<boolean>(false);
    const [showDeleteRecordsWarning, setShowDeleteRecordsWarning] = useState(false);
    const [sortCategory, setSortCategory] = useState<string>("");
    const [sortDirection, setSortDirection] = useState<string>("asc");
    const [selectedDisplayCategories, setSelectedDisplayCategories] = useState<string[]>([]);
    const { jwtToken } = useUser();
    const location = useLocation();
    const [currentPage, setCurrentPage] = useState(1);
    const { collectionId } = useParams();
    const queryClient = useQueryClient();
    const navigate = useNavigate();

    const hasSearchParams = new URLSearchParams(location.search).toString().length > 0;

    // Funkcja wyszukująca wartość dla danej kategorii
    const findValue = (artwork: any, categoryPath: any): string => {
        const parts = categoryPath.split(".");

        const searchCategory = (categories: any[], parts: string[]): string => {
            if (parts.length === 0) return "";
            const current = categories.find((cat) => cat.name === parts[0]);
            if (!current) return "";
            if (parts.length === 1) {
                return current.value || "";
            }
            if (current.subcategories && current.subcategories.length > 0) {
                return searchCategory(current.subcategories, parts.slice(1));
            }
            return "";
        };

        return searchCategory(artwork.categories, parts);
    };


    const {
        data: artworkData,
        isLoading: isLoadingArtworks,
        isFetching: isFetchingArtworks,
    } = useQuery({
        queryKey: [
            "artwork",
            collectionId,
            currentPage,
            location.search,
            sortCategory,
            sortDirection,
        ],
        queryFn: () =>
            getArtworksForCollectionPage(
                collectionId as string,
                currentPage,
                pageSize,
                `${sortCategory}-${sortDirection}`,
                new URLSearchParams(location.search).get("searchText"),
                Object.fromEntries(new URLSearchParams(location.search).entries())
            ),
        enabled: !!collectionId,
        keepPreviousData: false,
    });


    const { data: collectionData } = useQuery({
        queryKey: [collectionId],
        enabled: !!collectionId,
        queryFn: () => getCollection(collectionId as string),
    });

    const { data: categoriesData } = useQuery({
        queryKey: ["allCategories", collectionId],
        queryFn: () => getAllCategories(collectionId as string),
        enabled: !!collectionId,
    });


    // Przygotowanie opcji – lista wszyskich kategorii
    const categoryOptions: Option[] =
        categoriesData?.categories?.map((cat: string) => ({
            value: cat,
            label: cat,
        })) || [];


    // Ustaw domyślnie wybraną kategorię sortowania na pierwszą kategorię z listy (jeśli istnieje)
    useEffect(() => {
        if (categoriesData && categoriesData.categories && categoriesData.categories.length > 0 && !sortCategory) {
            setSortCategory(categoriesData.categories[0]);
        }
    }, [categoriesData, sortCategory]);

    // Ustaw domyślnie pierwsze 3 kategorie, jeśli jeszcze nie wybrano żadnych
    useEffect(() => {
        if (categoriesData && categoriesData.categories && selectedDisplayCategories.length === 0) {
            // Wyciągamy pierwsze 3 kategorie z listy
            setSelectedDisplayCategories(categoriesData.categories.slice(0, 3));
        }
    }, [categoriesData, selectedDisplayCategories]);

    const selectAll = () => {
        const newSelection = artworkData?.artworks.reduce((acc: any, artwork: any) => {
            acc[artwork._id] = true;
            return acc;
        }, {});
        setSelectedArtworks(newSelection);
    };

    const deselectAll = () => {
        setSelectedArtworks({});
    };

    const handleCheck = (id: string) => {
        setSelectedArtworks((prev) => ({ ...prev, [id]: !prev[id] }));
    };

    const deleteArtworksMutation = useMutation(
        () =>
            deleteArtworks(
                Object.keys(selectedArtworks).filter((id) => selectedArtworks[id]),
                jwtToken as string
            ),
        {
            onSuccess: () => {
                queryClient.invalidateQueries("artwork");
                setShowDeleteRecordsWarning((prev) => !prev);
                deselectAll();
            },
        }
    );

    if (!artworkData || !collectionData) {
        return (
            <div data-testid="loading-page-container">
                <LoadingPage />
            </div>
        );
    }


    type Option = {
        value: string;
        label: string;
    };

    // Dodanie opcji specjalnych na początku listy
    const customOptions = [
        { value: "select_all", label: "Zaznacz wszystkie" },
        { value: "deselect_all", label: "Odznacz wszystkie" },
        ...categoryOptions,
    ];

    const formatOptionLabel = (option: Option, { context }: { context: string }) => {
        if (context === "menu") {
            if (option.value === "select_all" || option.value === "deselect_all") {
                return (
                    <div
                        className="text-gray-500 dark:text-gray-300 underline"
                    >
                        {option.label}
                    </div>
                );
            }
        }
        return option.label;
    };

    return (
        <><div data-testid="loaded-artwork-page-container">
            <Navbar />
            {showDeleteRecordsWarning && (
                <WarningPopup
                    onClose={() => setShowDeleteRecordsWarning(false)}
                    deleteSelected={() => deleteArtworksMutation.mutate()}
                    warningMessage={"Czy na pewno chcesz usunąć zaznaczone rekordy?"}
                />
            )}
            <div className="flex flex-col w-full items-center bg-gray-50 dark:bg-gray-900 p-2 sm:p-4">
                <div className="flex flex-col max-w-screen-xl w-full lg:px-6">
                    <Navigation />
                    <div
                        data-testid="collection-name-and-description-container"
                        className="flex flex-row mb-4 mt-2"
                    >
                        <div className="flex flex-col w-full">
                            <h2 className="text-4xl font-bold text-gray-800 dark:text-white mb-1">
                                {collectionData?.name}
                            </h2>
                            <p className="text-xl text-gray-600 dark:text-gray-300">
                                {collectionData?.description}
                            </p>
                        </div>
                        {/* Edytuj kolekcję */}
                        <div>
                            <button
                                disabled={jwtToken ? false : true}
                                className={
                                    jwtToken
                                        ? "text-sm font-semibold h-fit mr-4 flex items-center"
                                        : "text-sm font-semibold h-fit mr-4 bg-gray-100 hover:bg-gray-100 flex items-center"
                                }
                                onClick={() =>  navigate(`/collections/${collectionId}/edit`, {
                                    state: {
                                        collectionId: collectionId,
                                        mode: 'edit',
                                        name: collectionData?.name,
                                        description: collectionData?.description,
                                        categories: collectionData?.categories
                                    }
                                })}
                            >
                                <EditIcon/>
                                <p className="ml-1">Edytuj</p>
                            </button>

                        </div>
                    </div>
                    {collectionId && <SearchComponent collectionId={collectionId} />}
                    <div className="flex w-full md:w-auto">
                        <div className="flex flex-1 space-x-2">
                            <button
                                disabled={jwtToken ? false : true}
                                className={`flex items-center justify-center dark:text-white hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 px-4 py-2 dark:focus:ring-primary-800 font-semibold text-white ${
                                    jwtToken
                                        ? "bg-gray-800 hover:bg-gray-700 border-gray-800"
                                        : "bg-gray-600 hover:bg-gray-600 border-gray-800"
                                }`}
                                type="button"
                                onClick={() => navigate(`/collections/${collectionId}/create-artwork`)}
                            >
                                <span className="mr-2 text-white dark:text-gray-400">
                                    <PlusIcon />
                                </span>
                                Nowy rekord
                            </button>
                            <button
                                className="flex items-center justify-center dark:text-white hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 font-medium px-4 py-2 dark:focus:ring-primary-800 font-semibold text-white bg-gray-800 hover:bg-gray-700 border-gray-800"
                                type="button"
                                onClick={async () => {
                                    setShowExportOptions((prev) => !prev);
                                }}
                            >
                                <span className="text-white dark:text-gray-400">
                                    <FileExportIcon />
                                </span>
                                Eksportuj plik
                            </button>
                            <button
                                disabled={jwtToken ? false : true}
                                className={`flex items-center justify-center dark:text-white hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 px-4 py-2 dark:focus:ring-primary-800 font-semibold text-white ${
                                    jwtToken
                                        ? "bg-gray-800 hover:bg-gray-700 border-gray-800"
                                        : "bg-gray-600 hover:bg-gray-600 border-gray-800"
                                }`}
                                type="button"
                                onClick={() => setShowImportOptions((prev) => !prev)}
                            >
                                <span className="text-white dark:text-gray-400">
                                    <FileImportIcon />
                                </span>
                                Importuj plik
                            </button>
                            <button
                                className="flex items-center justify-center dark:text-white hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 px-4 py-2 dark:focus:ring-primary-800 font-semibold text-white bg-gray-800 hover:bg-gray-700 border-gray-800"
                                type="button"
                                onClick={selectAll}
                            >
                                Zaznacz wszystkie
                            </button>
                            <button
                                className="flex items-center justify-center dark:text-white hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 px-4 py-2 dark:focus:ring-primary-800 font-semibold text-white bg-gray-800 hover:bg-gray-700 border-gray-800"
                                type="button"
                                onClick={deselectAll}
                            >
                                Odznacz wszystkie
                            </button>
                            <button
                                disabled={
                                    !(jwtToken && !Object.values(selectedArtworks).every((value) => value === false))
                                }
                                className={`flex items-center justify-center dark:text-white hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 px-4 py-2 dark:focus:ring-primary-800 font-semibold text-white ${
                                    jwtToken && !Object.values(selectedArtworks).every((value) => value === false)
                                        ? "bg-gray-800 hover:bg-gray-700 border-gray-800"
                                        : "bg-gray-600 hover:bg-gray-600 border-gray-800"
                                }`}
                                type="button"
                                onClick={() => {
                                    setShowDeleteRecordsWarning((prev) => !prev);
                                }}
                            >
                                Usuń zaznaczone
                            </button>
                        </div>
                    </div>
                    {showImportOptions && <ImportOptions onClose={() => setShowImportOptions(false)} collectionData={collectionData}/>}
                    {showExportOptions && <ExportOptions onClose={() => setShowExportOptions(false)} selectedArtworks={selectedArtworks} initialFilename={`${collectionData?.name}.xlsx`} />}
                    <div className="flex w-full md:w-auto pt-4 flex-row items-center text-sm">
                        <p className="pr-2">Wyświetlane kategorie:</p>
                        <DisplayCategoriesSelect
                            selectedDisplayCategories={selectedDisplayCategories}
                            setSelectedDisplayCategories={setSelectedDisplayCategories}
                            categoryOptions={categoryOptions} // wcześniej zdefiniowana tablica opcji
                            customOptions={customOptions}     // wcześniej zdefiniowana tablica z opcjami specjalnymi i zwykłymi
                            formatOptionLabel={formatOptionLabel} // funkcja wyróżniająca opcje specjalne
                        />
                        <p className="pl-2 pr-2">Sortuj według:</p>
                        {categoryOptions && (
                            <SortOptions
                            options={categoryOptions}
                            sortCategory={sortCategory}
                            sortDirection={sortDirection}
                            onSelectCategory={setSortCategory}
                            onSelectDirection={setSortDirection}
                            setCurrentPage={setCurrentPage}
                            />
                        )}
                    </div>
                </div>
            </div>

            <div className="flex flex-row">
                <div className="flex mx-auto flex-1 justify-end w-full"></div>
                <div data-testid="artworks-listed" className="w-full flex-2 lg:px-6 max-w-screen-xl">
                    <ArtworksList
                        artworksData={artworkData}
                        collectionId={collectionId as string}
                        isLoading={isLoadingArtworks}
                        isFetching={isFetchingArtworks}
                        hasSearchParams={hasSearchParams}
                        selectedDisplayCategories={selectedDisplayCategories}
                        selectedArtworks={selectedArtworks}
                        onToggleSelect={handleCheck}
                        findValue={findValue}
                        jwtToken={jwtToken}
                    />
                </div>

                <div className="mx-auto w-full flex-1"></div>
            </div>
            <div className="flex justify-center mb-2">
                <Pagination
                    currentPage={currentPage}
                    totalPages={Math.ceil(artworkData.total / pageSize)}
                    setCurrentPage={(page) => setCurrentPage(page)}
                    onPageChange={deselectAll}
                />
            </div>
        </div>
        </>
    );
};

export default ArtworksListPage;
