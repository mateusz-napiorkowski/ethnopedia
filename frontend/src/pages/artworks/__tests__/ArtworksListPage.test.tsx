import '@testing-library/jest-dom';
import { render, waitFor } from '@testing-library/react'
import ArtworksListPage from "../ArtworksListPage"
import { QueryClient, QueryClientProvider } from "react-query";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { UserContext } from '../../../providers/UserProvider';
import {jwtToken, loggedInUserContextProps, collectionData, artworkIds, artworkTitles, artworksData, artworksDataSecondPage} from './utils/consts'

const mockUseNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom') as any,
  useNavigate: () => mockUseNavigate,
}));

const mockGetArtworksForPage = jest.fn()
const mockDeleteArtworks = jest.fn()
jest.mock('../../../api/artworks', () => ({
    getArtworksForPage: () => mockGetArtworksForPage,
    deleteArtworks: (artworkId: string, jwtToken: string) => mockDeleteArtworks(artworkId, jwtToken)
}))

const mockGetCollection = jest.fn()
jest.mock('../../../api/collections', () => ({
    getCollection: () => mockGetCollection
}))

const mockGetAllCategories = jest.fn()
jest.mock('../../../api/categories', () => ({
    getAllCategories: () => mockGetAllCategories
}))

const queryClient = new QueryClient();
const user = userEvent.setup()

const renderPage = (
    queryClient: QueryClient, 
    userContextProps: any = {
        isUserLoggedIn: false,
        firstName: "",
        userId: "",
        jwtToken: undefined,
        setUserData: jest.fn()
    },
    collectionId: string = collectionData._id,
    pageSize = 3
    ) => {
        return render(
            <UserContext.Provider value={ userContextProps }>
                <QueryClientProvider client={queryClient}>
                    <MemoryRouter initialEntries={[`/collections/${collectionId}/artworks/`]}>
                        <Routes>
                            <Route path="/collections/:collectionId/artworks/" element={<ArtworksListPage pageSize={pageSize}/>}/>
                        </Routes>  
                    </MemoryRouter>
                </QueryClientProvider>    
            </UserContext.Provider>
        );
};

describe("ArtworksListPage tests", () => {
    beforeEach(() => {
        jest.clearAllMocks();
        queryClient.clear();
    });

    it("should render loading state", () => {     
        const {queryByTestId} = renderPage(queryClient)
        
        expect(queryByTestId("loading-page-container")).toBeInTheDocument()
        expect(queryByTestId("loaded-artwork-page-container")).not.toBeInTheDocument()
    })

    it("should render artworks list component after data is fetched from API, collection name and description should be displayed", async () => {
        mockGetArtworksForPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByTestId, queryByTestId} = renderPage(queryClient)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))

        expect(queryByTestId("loading-page-container")).not.toBeInTheDocument()
        expect(getByTestId("loaded-artwork-page-container")).toBeInTheDocument()

        await waitFor(() => expect(getByTestId('collection-name-and-description-container')).toMatchSnapshot())
    })

    it("should have edit collection, add record, import file, and delete selected buttons disabled when user is not logged in", async () => {
        mockGetArtworksForPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByRole, getByTestId} = renderPage(queryClient)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        const editCollectionButton = getByRole("button", {name: /edytuj/i})
        const addNewRecordButton = getByRole("button", {name: /nowy rekord/i})
        const importFileButton = getByRole("button", {name: /importuj dane/i})
        const deleteSelectedButton = getByRole("button", {name: /usuń zaznaczone/i})

        expect(editCollectionButton).toBeDisabled()
        expect(addNewRecordButton).toBeDisabled()
        expect(importFileButton).toBeDisabled()
        expect(deleteSelectedButton).toBeDisabled()
    })

    it("should have edit collection, add record and import file buttons enabled when user is logged in, delete selected button should stay disabled when no artworks are selected", async () => {
        mockGetArtworksForPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByRole, getByTestId} = renderPage(queryClient, loggedInUserContextProps)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        const editCollectionButton = getByRole("button", {name: /edytuj/i})
        const addNewRecordButton = getByRole("button", {name: /nowy rekord/i})
        const importFileButton = getByRole("button", {name: /importuj dane/i})
        const deleteSelectedButton = getByRole("button", {name: /usuń zaznaczone/i})

        expect(editCollectionButton).not.toBeDisabled()
        expect(addNewRecordButton).not.toBeDisabled()
        expect(importFileButton).not.toBeDisabled()
        expect(deleteSelectedButton).toBeDisabled()
    })
 
    it("should navigate to edit collection page when enabled edit collection button is clicked and pass state with current collection data", async () => {
        mockGetArtworksForPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByRole, getByTestId} = renderPage(queryClient, loggedInUserContextProps)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        const addNewRecordButton = getByRole("button", {name: /edytuj/i})
        await user.click(addNewRecordButton)

        expect(mockUseNavigate).toHaveBeenCalledWith(`/collections/${collectionData._id}/edit`,
            {"state":
                {
                    "categories": collectionData.categories,
                    "collectionId": collectionData._id,
                    "description": collectionData.description,
                    "mode": "edit",
                    "name": collectionData.name}
            }
        )
    })

    it("should navigate to add record page when enabled add record button is clicked", async () => {
        mockGetArtworksForPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByRole, getByTestId} = renderPage(queryClient, loggedInUserContextProps)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        const addNewRecordButton = getByRole("button", {name: /nowy rekord/i})
        await user.click(addNewRecordButton)

        expect(mockUseNavigate).toHaveBeenCalledWith(`/collections/${collectionData._id}/create-artwork`)
    })

    it("should navigate to import to existing collection page when enabled import data button is clicked", async () => {
        mockGetArtworksForPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByRole, getByTestId, getByText} = renderPage(queryClient, loggedInUserContextProps)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        const importFileButton = getByRole("button", {name: /importuj dane/i})
        await user.click(importFileButton)

        expect(mockUseNavigate).toHaveBeenCalledWith(`/collections/${collectionData._id}/import-data`)
    })

    it("should navigate to export data page and pass correct state when export file button is clicked", async () => {
        mockGetArtworksForPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByRole, getByTestId} = renderPage(queryClient, loggedInUserContextProps)
        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        
        const exportFileButton = getByRole("button", {name: /eksportuj dane/i})
        
        await user.click(exportFileButton)
        expect(mockUseNavigate).toHaveBeenCalledWith(
            `/collections/${collectionData._id}/export-data`,
            {
                state: {
                    "initialArchiveFilename": collectionData.name,
                    "initialFilename": collectionData.name,
                    "searchParams": "",
                    "selectedArtworks": {}
                }
            }
        )
       
    })

    it("should expand display categories select menu when it is clicked", async () => {
        mockGetArtworksForPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByRole, getByTestId, getByLabelText} = renderPage(queryClient, loggedInUserContextProps)
        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        
        const expandDisplayCategoriesSelectElement = getByLabelText("open/close-display-categories-select")

        await user.click(expandDisplayCategoriesSelectElement)
        expect(getByTestId("DisplayCategoriesSelectExpanded")).toBeInTheDocument()
    })

    it("should navigate to appropriate artwork page when entry from artworks list is clicked", async () => {
        mockGetArtworksForPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByTestId} = renderPage(queryClient)

        await waitFor(() => getByTestId('artworks-listed'))

        await user.click(getByTestId(artworkIds[0]))
        
        expect(mockUseNavigate).toHaveBeenCalledWith(`/collections/${collectionData._id}/artworks/${artworkIds[0]}`)
    })

    it("should have delete selected button enabled if any artworks are selected", async () => {
        mockGetArtworksForPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByRole, getByTestId} = renderPage(queryClient, loggedInUserContextProps)
        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        const deleteSelectedButton = getByRole("button", {name: /usuń zaznaczone/i})
        const exampleArtworkCheckbox = getByTestId(`${artworkIds[0]}-checkbox`)

        await user.click(exampleArtworkCheckbox)

        expect(deleteSelectedButton).not.toBeDisabled()
    })

    it("should refetch artworks list when artwork page is changed", async () => {
        mockGetArtworksForPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        mockGetAllCategories.mockReturnValueOnce({categories: ["Tytuł", "Tytuł.Podtytuł", "Artyści", "Rok"]})
        const {getByTestId} = renderPage(queryClient)
        await waitFor(() => getByTestId('loaded-artwork-page-container'))

        await waitFor(() => expect(getByTestId('artworks-listed')).toMatchSnapshot("page 1"))

        mockGetArtworksForPage.mockReturnValueOnce(artworksDataSecondPage)
        await user.click(getByTestId("page-2"))

        await waitFor(() => expect(getByTestId('artworks-listed')).toMatchSnapshot("page 2"))
    })

    it.each([
            {
                clickSequence: [],
                clickSequenceWithTitles: [],
                expectedChecked: []
            },
            {
                clickSequence: [artworkIds[1], artworkIds[2], artworkIds[0], artworkIds[2], artworkIds[1], artworkIds[2]],
                clickSequenceWithTitles: [artworkTitles[1], artworkTitles[2], artworkTitles[0], artworkTitles[2], artworkTitles[1], artworkTitles[2]],
                expectedChecked: [artworkIds[2], artworkIds[0]]
            },
            {
                clickSequence: [artworkIds[1], artworkIds[0], artworkIds[2], artworkIds[0], "Odznacz wszystkie"],
                clickSequenceWithTitles: [artworkTitles[1], artworkTitles[0], artworkTitles[2], artworkIds[0], "Odznacz wszystkie"],
                expectedChecked: []
            },
            {
                clickSequence: [artworkIds[1], "Zaznacz wszystkie"],
                clickSequenceWithTitles: [artworkTitles[1], "Zaznacz wszystkie"],
                expectedChecked: [artworkIds[0], artworkIds[1], artworkIds[2]]
            },
          ])('should have correct artworks checked and call deleteArtwork with correct args for checkboxes/buttons clicked in sequence: $clickSequenceWithTitles', async ({clickSequence, expectedChecked}) => {
            const expectedUnchecked = artworkIds.filter(category => !expectedChecked.includes(category))
            mockGetArtworksForPage.mockReturnValue(artworksData)
            mockGetCollection.mockReturnValue(collectionData)
            const {getByTestId, getByRole, getByLabelText, getByText} = renderPage(queryClient, loggedInUserContextProps)
            
            await waitFor(() => getByTestId('artworks-listed'))
            for(const value of clickSequence) {
                const elementToClick = (value === "Zaznacz wszystkie" || value === "Odznacz wszystkie")
                    ? getByText(value) : getByTestId(`${value}-checkbox`)
                await user.click(elementToClick)
            }
            
            expectedChecked.forEach(value => {
                expect(getByTestId(`${value}-checkbox`)).toBeChecked();
            })
            expectedUnchecked.forEach(value => {
                expect(getByTestId(`${value}-checkbox`)).not.toBeChecked();
            })

            const deleteSelectedButton = getByRole("button", {name: /usuń zaznaczone/i}) as HTMLInputElement
            if (!deleteSelectedButton.disabled) {
                await user.click(deleteSelectedButton)
                await user.click(getByLabelText("confirm"))
                await waitFor(() => expect(mockDeleteArtworks).toHaveBeenCalledWith(expectedChecked, jwtToken))
                expectedChecked.forEach(value => {
                    expect(getByTestId(`${value}-checkbox`)).not.toBeChecked();
                })
            }
        })
})