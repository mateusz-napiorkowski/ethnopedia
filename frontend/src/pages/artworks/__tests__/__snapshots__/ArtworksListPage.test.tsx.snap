// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtworksListPage tests should refetch artworks list when artwork page is changed: page 1 1`] = `
<div
  class="w-full flex-2 lg:px-6 max-w-screen-xl"
  data-testid="artworks-listed"
>
  <div
    class="px-4 max-w-screen-xl py-4 bg-white dark:bg-gray-800 shadow-md w-full rounded-lg mb-4 border border-gray-300 dark:border-gray-600 cursor-pointer"
    data-testid="6752ddca46e3ca48231024dc"
  >
    <div
      class="flex flex-row"
    >
      <span
        class="mr-4 flex items-center"
      >
        <input
          data-testid="6752ddca46e3ca48231024dc-checkbox"
          type="checkbox"
        />
      </span>
      <div>
        <div
          class="text-lg text-gray-800 dark:text-white"
        >
          <p
            class="text-gray-400 inline"
          >
            Tytuł
            : 
          </p>
          Jakym jechoł koło dworu
        </div>
      </div>
    </div>
  </div>
  <div
    class="px-4 max-w-screen-xl py-4 bg-white dark:bg-gray-800 shadow-md w-full rounded-lg mb-4 border border-gray-300 dark:border-gray-600 cursor-pointer"
    data-testid="6752ddca46e3ca48231024aa"
  >
    <div
      class="flex flex-row"
    >
      <span
        class="mr-4 flex items-center"
      >
        <input
          data-testid="6752ddca46e3ca48231024aa-checkbox"
          type="checkbox"
        />
      </span>
      <div>
        <div
          class="text-lg text-gray-800 dark:text-white"
        >
          <p
            class="text-gray-400 inline"
          >
            Tytuł
            : 
          </p>
          Ojcze, ojcze, kup mi kónia
        </div>
      </div>
    </div>
  </div>
  <div
    class="px-4 max-w-screen-xl py-4 bg-white dark:bg-gray-800 shadow-md w-full rounded-lg mb-4 border border-gray-300 dark:border-gray-600 cursor-pointer"
    data-testid="6752ddca46e3ca48231024bb"
  >
    <div
      class="flex flex-row"
    >
      <span
        class="mr-4 flex items-center"
      >
        <input
          data-testid="6752ddca46e3ca48231024bb-checkbox"
          type="checkbox"
        />
      </span>
      <div>
        <div
          class="text-lg text-gray-800 dark:text-white"
        >
          <p
            class="text-gray-400 inline"
          >
            Tytuł
            : 
          </p>
          Piękna jagoda
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ArtworksListPage tests should refetch artworks list when artwork page is changed: page 2 1`] = `
<div
  class="w-full flex-2 lg:px-6 max-w-screen-xl"
  data-testid="artworks-listed"
>
  <div
    class="px-4 max-w-screen-xl py-4 bg-white dark:bg-gray-800 shadow-md w-full rounded-lg mb-4 border border-gray-300 dark:border-gray-600 cursor-pointer"
    data-testid="6752aaaa46e3ca48231024dc"
  >
    <div
      class="flex flex-row"
    >
      <span
        class="mr-4 flex items-center"
      >
        <input
          data-testid="6752aaaa46e3ca48231024dc-checkbox"
          type="checkbox"
        />
      </span>
      <div>
        <div
          class="text-lg text-gray-800 dark:text-white"
        >
          <p
            class="text-gray-400 inline"
          >
            Tytuł
            : 
          </p>
          W górach, nad rzeką
        </div>
      </div>
    </div>
  </div>
  <div
    class="px-4 max-w-screen-xl py-4 bg-white dark:bg-gray-800 shadow-md w-full rounded-lg mb-4 border border-gray-300 dark:border-gray-600 cursor-pointer"
    data-testid="6752aaaa46e5db48231024dc"
  >
    <div
      class="flex flex-row"
    >
      <span
        class="mr-4 flex items-center"
      >
        <input
          data-testid="6752aaaa46e5db48231024dc-checkbox"
          type="checkbox"
        />
      </span>
      <div>
        <div
          class="text-lg text-gray-800 dark:text-white"
        >
          <p
            class="text-gray-400 inline"
          >
            Tytuł
            : 
          </p>
          Skrzypce na polu
        </div>
      </div>
    </div>
  </div>
  <div
    class="px-4 max-w-screen-xl py-4 bg-white dark:bg-gray-800 shadow-md w-full rounded-lg mb-4 border border-gray-300 dark:border-gray-600 cursor-pointer"
    data-testid="1234aaca46e3ca48231024bb"
  >
    <div
      class="flex flex-row"
    >
      <span
        class="mr-4 flex items-center"
      >
        <input
          data-testid="1234aaca46e3ca48231024bb-checkbox"
          type="checkbox"
        />
      </span>
      <div>
        <div
          class="text-lg text-gray-800 dark:text-white"
        >
          <p
            class="text-gray-400 inline"
          >
            Tytuł
            : 
          </p>
          Pragnom uoczka, pragnom
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ArtworksListPage tests should render artworks list component after data is fetched from API, collection name and description should be displayed 1`] = `
<div
  class="flex flex-row mb-4 mt-2"
  data-testid="collection-name-and-description-container"
>
  <div
    class="flex flex-col w-full"
  >
    <h2
      class="text-4xl font-bold text-gray-800 dark:text-white mb-1"
    >
      example collection
    </h2>
    <p
      class="text-xl text-gray-600 dark:text-gray-300"
    >
      example collection description
    </p>
  </div>
</div>
`;
